plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -311,89 +396,192 @@ func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
 /*
  * Gather given Metric and emit any error
  */
-func (c *CloudWatch) gatherMetric(
-	acc telegraf.Accumulator,
-	metric *cloudwatch.Metric,
-) error {
-	params := c.getStatisticsInput(metric)
-	resp, err := c.client.GetMetricStatistics(params)
+func (c *CloudWatch) gatherMetrics(
+	params *cloudwatch.GetMetricDataInput,
+) ([]*cloudwatch.MetricDataResult, error) {
+	resp, err := c.client.GetMetricData(params)
 	if err != nil {
-		return err
+		return nil, errors.New("Failed to get metric data - " + err.Error())
 	}
 
-	for _, point := range resp.Datapoints {
+	// todo: determine if we need to handle pagination.
+	if resp.NextToken != nil {
+		log.Println("[inputs.cloudwatch] W! UNHANDLED PAGINATED RESULTS!!", *resp.NextToken)
+	}
+	return resp.MetricDataResults, nil
+}
+
+func (c *CloudWatch) aggregateMetrics(
+	acc telegraf.Accumulator,
+	metricDataResults []*cloudwatch.MetricDataResult,
+) error {
+	namespace := formatMeasurement(c.Namespace)
+
+	for _, query := range c.queries {
 		tags := map[string]string{
 			"region": c.Region,
-			"unit":   snakeCase(*point.Unit),
 		}
 
-		for _, d := range metric.Dimensions {
-			tags[snakeCase(*d.Name)] = *d.Value
+		type metric struct {
+			fields map[string]interface{}
+			tags   map[string]string
 		}
+		thing := map[time.Time]metric{}
+		for _, result := range metricDataResults {
+			if nameMatch(query.id, *result.Id) {
+				for _, dimension := range query.metric.Dimensions {
+					tags[snakeCase(*dimension.Name)] = *dimension.Value
+				}
+				if len(result.Timestamps) != len(result.Values) {
+					log.Println("[inputs.cloudwatch] W! MISMATCHED TIMESTAMP/VALUE LENGTH!!", len(result.Timestamps), len(result.Values))
```

Comment:
As an aside, when would this happen?

Reply:
Not sure, figured I'd log and be told when it did... I suppose a `todo: check aws code to ensure timestamps and values always match up` would have served the same purpose

Reply:
https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_MetricDataResult.html